c10/util/typeid.h

```diff
@@ -442,29 +458,115 @@ class C10_API TypeMeta final {
 #endif
   }
 
- private:
-  const detail::TypeMetaData* data_;
+  /**
+  * convert ScalarType enum values to TypeMeta handles
+  */
+  static inline caffe2::TypeMeta fromScalarType(ScalarType scalar_type) {
+    const size_t index = static_cast<uint16_t>(scalar_type);
+    if (C10_LIKELY(index < NumScalarTypes)) {
+      return TypeMeta(index);
+    }
+    error_unrecognized_scalartype(scalar_type);
+    return TypeMeta(); // avoids "control reaches end of non-void function"
+  }
 
+  /**
+   * convert TypeMeta handles to ScalarType enum values
+   */
+  inline ScalarType toScalarType() {
+    if (C10_LIKELY(isScalarType())) {
+      return static_cast<ScalarType>(index_);
+    }
+    error_unsupported_typemeta(*this);
+    return ScalarType::Undefined; // avoids "control reaches end of non-void function"
+  }
+
+private:
+  static void error_unrecognized_scalartype(ScalarType scalar_type) {
+    TORCH_CHECK(false, "Unrecognized Scalartype ", scalar_type, " (please report this error)");
+  }
+  static void error_unsupported_typemeta(caffe2::TypeMeta dtype) {
+    TORCH_CHECK(false, "Unsupported TypeMeta in ATen: ", dtype, " (please report this error)");
+  }
+
+  // preallocated TypeMetaData instances for ScalarType types
+  static const detail::TypeMetaData& scalarTypeMetaData(uint16_t index);
+
+  // TypeMetaData instances for non-ScalarType types...
+  static std::vector<detail::TypeMetaData>& nonScalarTypeMetaDatas();
+
+  // ...added through here
+  static std::mutex appendMutex_;
   template <class T>
-  C10_API static const detail::TypeMetaData* _typeMetaDataInstance() noexcept;
+  static uint16_t addTypeMetaDataInstance() {
+    auto typeId = TypeIdentifier::Get<T>();
+    auto typeName = c10::util::get_fully_qualified_type_name<T>();
+    std::vector<detail::TypeMetaData>& instances = nonScalarTypeMetaDatas();
+    // concurrency situation:
+    // the static init for each _typeMetaDataInstance<T> is synchronized,
+    // so we don't need to worry about the same type being added twice.
+    // But different threads might be adding different Ts concurrently.
+    // So while we don't need to test for the presence of T before
+    // appending, we do need to serialize the appends themselves.
```

Comment:
Nuh uh, you can't do this. If you have a reader access to vector in one thread, and a writer access to vector in another, you need synchronization on both ends. Writer synchronization isn't enough.

Reply:
Which, for obvious reasons, is going to be a big performance problem.

Reply:
cc @smessmer

Reply:
The big hazard is `instances` reallocating the buffer. So one possible way we could fix this is add a compile-time "maximum number of typeids" and then just allocate the buffer large enough. Then concurrent reads could be implemented in a lock free manner.

Reply:
Ugh yeah realized this with a sinking feeling while brushing my teeth this morning. I think I temporarily hallucinated eager initialization that didn't interleave with accesses... so TypeMetaData is 72 bytes, and a quick grep says there are currently no more than 71 CAFFE_KNOWN_TYPEs. 256 would be a little over 18k. Does that sound reasonable (in both senses)?